tls/s2n_client_hello.c

```diff
@@ -348,6 +355,16 @@ int s2n_client_hello_recv(struct s2n_connection *conn)
 
 int s2n_client_hello_send(struct s2n_connection *conn)
 {
+    const struct s2n_cipher_preferences *cipher_preferences;
+    GUARD(s2n_connection_get_cipher_preferences(conn, &cipher_preferences));
+
+    /* Check whether cipher preference supports TLS 1.3. If it doesn't,
+       our higest supported version is S2N_TLS12 */
+    if (!s2n_cipher_preference_is_tls13_required(cipher_preferences)) {
+        conn->client_protocol_version = MIN(conn->client_protocol_version, S2N_TLS12);
```

Comment:
Does this silently downgrade the version? Would you rather want to fail the handshake instead?

Reply:
This is for s2n clients using tls12 cipher suites. Although s2n *can* send client hello with TLS 1.3, it's implied that the client wants TLS 1.2 behavior, and we are ensure that 1.2 is negotiated as it's configuration is not safe for TLS 1.3 otherwise. If a client wants to negotiate 1.3, it should be using a TLS1.3 cipher preference